LuaUI/Widgets/gbc/Pathfinding.lua

```diff
@@ -0,0 +1,161 @@
+
+-- luacheck: read globals WG
+-- provides: read globals UpdateOneWorkerPathing UpdateOneJobPathing CleanPathing
+
+-- Global Build Command/Pathfinding: Responsible keeping track of which units
+-- can reach which jobs and vice versa.
+-- Gives a lot of false positives: Spring.RequestPath currently seems overly
+-- generous, and will often provide a path if it thinks the unit can get
+-- closer without necessarily being able to reach the target.
+-- XXX: State is currently still tightly coupled with core GBC state.
+
+local Echo               = Spring.Echo
+local spGetUnitDefID     = Spring.GetUnitDefID
+local spGetUnitPosition  = Spring.GetUnitPosition
+local spIsUnitAllied     = Spring.IsUnitAllied
+local spValidUnitID      = Spring.ValidUnitID
+local spUnitIsDead       = Spring.GetUnitIsDead
+local spRequestPath      = Spring.RequestPath
+
+local CMD_REPAIR         = CMD.REPAIR
+local CMD_RECLAIM        = CMD.RECLAIM
+
+local abs                = math.abs
+local sqrt               = math.sqrt
+
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+-- Helper Functions ------------------------------------------------------------
+--[[
+HOW THIS WORKS:
+	-- EXTERNAL API --
+	UpdateOneWorkerPathing()
+		Caches pathing info for one worker for every job in buildQueue. This is called
+		whenever a new unit enters our group, and adds the hash of any job that cannot be reached
+		and/or performed to 'includedBuilders[unitID].unreachable'. Does not do anything for commands targetting
+		units (ie repair, reclaim), since they may move and invalidate cached pathing info.
+	UpdateOneJobPathing()
+		Caches pathing info for one job for every worker in includedBuilders. This is called whenever a new
+		job is added to the queue, and does basically the same thing as UpdateOneWorkerPathing().
+	CleanPathing()
+		Performs garbage collection for unreachable caches, removing jobs that are no longer on the queue.
+	-- INTERNAL FUNCTIONS --
+	CanBuildThis()
+		Determines whether a given worker can perform a given job or not.
+	IsTargetReachable()
+		Checks pathing between a unit and destination, to determine if a worker
+		can reach a given build site.
+--]]
+
+--This function tells us if a unit can perform the job in question.
+local function CanBuildThis(cmdID, unitID)
+	local unitDefID = spGetUnitDefID(unitID)
+	local unitDef = UnitDefs[unitDefID]
+	if cmdID < 0 then -- for build jobs
+		local bcmd = abs(cmdID) -- abs the command ID to get the unitDefID that it refers to
```

Comment:
We know `< 0` so we can use the cheaper operator instead of invoking a function
```suggestion
		local bcmd = -cmdID
```